File: CivitasJava/src/civitas/Tablero.java
package civitas;

import java.util.ArrayList;

public class Tablero {
    private int numCasillaCarcel;
    private ArrayList<Casilla> casillas;
    private int porSalida;
    private boolean tieneJuez;
    
    public Tablero(int numCasillaCarcel){
        if(numCasillaCarcel >= 1){
            this.numCasillaCarcel = numCasillaCarcel;
        }
        
        else{
            this.numCasillaCarcel = 1;
        }
        
        this.casillas = new ArrayList<Casilla>();
        Casilla salida = new Casilla ("Salida");
        casillas.add(salida);
        
        porSalida=0;
        tieneJuez=false;        
    }
    
    private boolean correcto(){
        return((casillas.size() > numCasillaCarcel) && tieneJuez);
    }
    
    private boolean correcto(int numCasilla){
        return(correcto() && numCasilla >= 0 && numCasilla < casillas.size());
    }
    
     int getCarcel(){
        return numCasillaCarcel;
    }
    
    int getPorSalida(){
        int auxPorSalida = porSalida;
        
        if(porSalida>0){
            porSalida--;
        }
        
        return auxPorSalida;
    }
    
    
    void añadeCasilla(Casilla casilla){
        if(casillas.size() == numCasillaCarcel){
            Casilla carcel = new Casilla("Cárcel");
            casillas.add(carcel);
        }
        
        casillas.add(casilla);
        
        if(casillas.size() == numCasillaCarcel){
            Casilla carcel = new Casilla("Cárcel");
            casillas.add(carcel);
        }
    }
    
    void añadeJuez(){
        if(!tieneJuez){
            CasillaJuez juez = new CasillaJuez(numCasillaCarcel, "Juez");
            casillas.add(juez);
            tieneJuez=true;
        }
    }
    
    Casilla getCasilla(int numCasilla){
        Casilla casilla=null;
        if(correcto(numCasilla)){
            casilla = casillas.get(numCasilla);           
        }
        
        return casilla;
    }
    
    public ArrayList<Casilla> getCasillas(){
        return casillas;
    }
    
    int nuevaPosicion(int actual, int tirada){
        int nuevaPosicion = -1;

        if(correcto()){
            nuevaPosicion = actual + tirada;
            if (nuevaPosicion > casillas.size()){
                nuevaPosicion = nuevaPosicion%casillas.size();
                porSalida++;
            }

        }

        return nuevaPosicion;
    }
    
    int calcularTirada(int origen, int destino){
        int tirada = destino - origen;
        
        if(tirada < 0){
            tirada = tirada + casillas.size();           
            porSalida++;
        }
        
        return tirada;
    }
}

File: CivitasRuby/pkg/CivitasRuby-0.0.1/lib/tablero.rb
#encoding:utf-8

require_relative 'casilla'

module Civitas

  class Tablero
    def initialize (casilla_carcel)
      if(casilla_carcel >= 1)
        @num_casilla_carcel=casilla_carcel
      else
        @num_casilla_carcel=1
      end
      
      @por_salida=0
      @tiene_juez=false
      @casillas=[]

      salida=Casilla.new("Salida")
      @casillas << salida
    end

    def correcto
      result = (@casillas.length > @num_casilla_carcel) && (@tiene_juez)
      
      return result
    end
    
    def correcto_cas(num_casilla)
      
      result = (correcto) && (num_casilla >= 0 && num_casilla < @casillas.length)
      puts @casillas.length
      
      return result
    end
    
    attr_reader :casillas
    
    def aniade_casilla(casilla)
      if(@casillas.length == @num_casilla_carcel)
        carcel=Casilla.new("Carcel")
        @casillas << carcel
      end
      
      @casillas << casilla
      
      if(@casillas.length == @num_casilla_carcel)
        carcel=Casilla.new("Carcel")
        @casillas << carcel
      end
    end
    
    def aniade_juez
      if(@tiene_juez == false)
        juez=CasillaJuez.new(@num_casilla_carcel, "Juez")
        @casillas << juez
        @tiene_juez = true
      end
    end
    
    def get_casilla(num_casilla)
      casilla=nil
      
      if(correcto_cas(num_casilla))
        casilla=@casillas[num_casilla]
      end
      
      return casilla
    end
    
    def nueva_posicion(actual, tirada)
      posicion_nueva=-1
      
      if(correcto)
        posicion_nueva=actual+tirada
        if(posicion_nueva > @casillas.length-1)
          posicion_nueva=posicion_nueva%@casillas.length
          @por_salida += 1
        end
      end
      
      return posicion_nueva
    end
    
    def calcular_tirada(origen, destino)
      tirada=destino-origen
      
      if(tirada<0)
        tirada=tirada+@casillas.length
        @por_salida += 1
      end
      
      return tirada
    end
    
    def get_num_casilla_carcel
      return @num_casilla_carcel
    end
    
    def get_por_salida
        aux_por_salida=@por_salida

        if(@por_salida>0)
          @por_salida-=1
        end
        
        return aux_por_salida
    end
    
    
    private  :correcto ,:correcto_cas
    
    
  end 
end
